### PR TITLE
[codex] Fix GitHub extension search dispatch

### PR DESCRIPTION
**Pull request by serrrfirat**
## Summary

- Accept structured `github.search_issues` inputs such as `repo`, `repository`, `author`, `state`, and `type`, then build the GitHub search query inside the bundled WASM extension.
- Map GitHub WASM guest failures into useful host runtime error categories instead of collapsing everything to `operation_failed`.
- Initialize the default WASM runtime for local-dev and refresh materialized bundled extension assets during restore so installed GitHub extensions pick up updated bundled WASM.

## Stack

This PR is stacked on top of #4281 (`codex/fix-webui-manual-token-resume`). Review the diff against that branch.

## Root Cause

The GitHub extension only accepted a raw `query` string, while model calls can naturally produce structured search parameters. Those rejected inputs, plus GitHub/API failures, were surfaced as generic `operation_failed`. Already-installed bundled extensions could also keep using stale materialized WASM assets.

## Validation

- `cargo fmt`
- `cargo fmt --manifest-path crates/ironclaw_first_party_extensions/assets/github/wasm-src/Cargo.toml`
- `cargo test --manifest-path crates/ironclaw_first_party_extensions/assets/github/wasm-src/Cargo.toml`
- `cargo test -p ironclaw_host_runtime --test github_wasm_runtime_contract`
- `cargo test -p ironclaw_reborn_composition restore_refreshes_materialized_extension_assets_from_catalog --features libsql`
- `cargo test -p ironclaw_reborn_composition local_dev_github_installs_activates_and_reaches_auth_gate --features libsql`
- `cargo test -p ironclaw_reborn_composition local_dev_capability_port_restores_activated_github_extension_surface --features libsql`
- `git diff --check origin/codex/fix-webui-manual-token-resume..HEAD`

### CODE REVIEW

File: crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json [hunk 1]
```diff
@@ -36,5 +81,13 @@
       "description": "Sort order when sort is supplied."
     }
   },
-  "required": ["query"]
+  "anyOf": [
+    { "required": ["query"] },
+    { "required": ["repository"] },
+    { "required": ["owner", "repo"] },
+    { "required": ["repo"] },
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json]: **Medium - schema/runtime mismatch**

The schema makes `repo` alone a valid search input, but the guest treats `repo` without `owner` as an `owner/repo` shorthand and rejects values without a slash. A structured call like `{ "repo": "ironclaw" }` is therefore advertised as valid by the schema but fails before GitHub dispatch.

Fix: add conditional schema rules so `repo` must match `owner/repo` when `owner` is absent, and require `repo` whenever `owner` is present.

**Review by serrrfirat — COMMENTED** [review]
Multi-agent code review for #4282 (`--force`). Reviewed as a stacked PR against `codex/fix-webui-manual-token-resume`.

Reviewers completed: Security, Bugs, Performance/Concurrency, Tests, Conventions.

Result: COMMENT. No Critical/High findings. Medium findings remain around schema/runtime consistency, generic WASM adapter boundaries, and missing caller-facing coverage.

Notes: I filtered one security finding about refreshing host-bundled assets under the same manifest hash because this path restores app-shipped first-party bundled assets, not user-supplied extension code. The remaining findings below are actionable within this PR scope.

File: crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json [hunk 2]
```diff
@@ -10,6 +10,51 @@
       "maxLength": 512,
       "description": "GitHub issue search query. Use qualifiers such as repo:, org:, is:issue, is:pr, label:, state:, author:, assignee:, or involves:."
     },
+    "repository": {
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json:13]: **Low - unbounded schema input**

`repository` is the only new string search field without a length bound. Even though the WASM guest validates segments later, schema validation can admit a very large string first.

Fix: add `minLength` and `maxLength` for this shorthand, for example `minLength: 3` and `maxLength: 201` for two 100-character segments plus `/`.

File: crates/ironclaw_host_runtime/src/services/runtime_adapters.rs [hunk 3]
```diff
@@ -768,6 +774,22 @@ fn dispatch_error_for_runtime(
     }
 }
 
+fn wasm_guest_error_kind(error: &str) -> RuntimeDispatchErrorKind {
```
Comment (serrrfirat) [inline crates/ironclaw_host_runtime/src/services/runtime_adapters.rs]: **Medium - boundary leakage**

`wasm_guest_error_kind` lives in the generic host-runtime WASM adapter, but it now knows GitHub-specific guest errors such as `github_api_*` and `unsupported_github_capability`. That makes every extension's private error vocabulary a host-runtime concern.

Fix: have the GitHub guest emit runtime-agnostic error categories/codes, or translate GitHub-specific strings before they reach the generic host-runtime adapter.

File: crates/ironclaw_host_runtime/src/services/runtime_adapters.rs [hunk 4]
```diff
@@ -595,14 +595,20 @@ where
...
         )?;
+        if error == "AuthRequired" {
+            return Err(DispatchError::AuthRequired {
+                capability: request.capability_id.clone(),
+                required_secrets: Vec::new(),
+            });
+        }
         return Err(DispatchError::Wasm {
-            kind: RuntimeDispatchErrorKind::OperationFailed,
+            kind: wasm_guest_error_kind(&error),
```
Comment (serrrfirat) [inline crates/ironclaw_host_runtime/src/services/runtime_adapters.rs]: **Medium - missing dispatch-level error mapping tests**

The PR changes guest execution errors from generic `operation_failed` into `AuthRequired`, `InputEncode`, `Client`, `NetworkDenied`, `OutputTooLarge`, and `Executor` categories. Existing tests assert bundled WASM error strings, but not the final `RuntimeCapabilityOutcome` / dispatch failure kind produced by host runtime.

Fix: add a host-runtime contract test covering representative guest errors through `HostRuntimeServices`, including `AuthRequired`, `invalid_parameters`, `github_api_forbidden`, `github_api_rate_limited`, `github_api_body_limit`, `github_api_timeout`, and `github_api_egress_denied`.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 5]
```diff
@@ -127,7 +137,8 @@ fn operation_from_context(context: Option<&str>) -> Result<GitHubOperation, Stri
 }
 
 fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
-    validate_text(&params.query, "query", MAX_QUERY_LENGTH)?;
+    let query = search_query(&params)?;
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:142]: **Medium - structured search lacks host-runtime coverage**

The structured search path is covered by direct bundled-WASM execution, but the host-runtime GitHub contract still dispatches only raw `query` input. That leaves schema exposure, credential injection, runtime HTTP egress, and host invocation with structured fields uncovered together.

Fix: add a `HostRuntimeServices` test that invokes `github.search_issues` with structured `repo`/`author`/`state`/`type` input and asserts the expected GitHub request.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 6]
```diff
@@ -156,6 +167,73 @@ fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
...
+    let mut parts = Vec::new();
+    if let Some(query) = params
+        .query
+        .as_deref()
+        .map(str::trim)
+        .filter(|query| !query.is_empty())
+    {
+        parts.push(query.to_string());
+    }
+    if let Some(repository) = params.repository.as_deref() {
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:185]: **Low - structured rejection cases are under-tested**

`search_query` adds new rejection paths for malformed repository inputs, owner without repo, invalid qualifier values, invalid state, and invalid type. The adjacent parameter rejection test still mostly covers the old raw-query/page/limit/sort cases.

Fix: add a focused guest test for invalid structured search fields: owner without repo, malformed `repository` / `repo` shorthand, whitespace author, invalid state, and invalid type.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 7]
```diff
@@ -156,6 +167,73 @@ fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
...
+        let (owner, repo) = repository
+            .split_once('/')
+            .ok_or_else(|| "invalid_repository".to_string())?;
+        validate_repo(owner, repo)?;
+        parts.push(format!("repo:{owner}/{repo}"));
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:190]: **[BUG-1] Duplicate `repo:` qualifier** — `repository` pushes `repo:{owner}/{repo}` here, but the `match (params.owner.as_deref(), params.repo.as_deref())` block below (lines 187–200) may also push a second `repo:` qualifier if `owner`+`repo` or a bare `repo` shorthand is supplied. No mutual-exclusion guard exists.

Fix — add at the top of `search_query`:
```rust
if params.repository.is_some() && (params.owner.is_some() || params.repo.is_some()) {
    return Err("invalid_repository".to_string());
}
```

**Review by serrrfirat — COMMENTED** [review]
## Code Review — #4282: Fix GitHub extension search dispatch

**Reviewers:** Security · Bugs · Performance · Tests · Conventions  
**Findings:** 15 (2 High, 7 Medium, 4 Low, 2 filtered false-positives)

---

### 🔴 High — Bugs

**[BUG-1] Duplicate `repo:` qualifier when `repository` and `owner`/`repo` both provided**  
`lib.rs:170–200` — `search_query()` processes `repository` first, then `owner`/`repo`, with no mutual-exclusion guard. Supplying both emits two `repo:` qualifiers (`repo:a/b repo:c/d`). GitHub silently picks one, producing wrong/empty results with no error returned.  
**Fix:** Early guard before processing:
```rust
if params.repository.is_some() && (params.owner.is_some() || params.repo.is_some()) {
    return Err("invalid_repository".to_string());
}
```

**[BUG-2] `anyOf: [{"required": ["repo"]}]` contradicts runtime: bare `repo` always errors**  
`schema:88` + `lib.rs:192–199` — Schema advertises `repo` alone as valid. The `(None, Some(repo))` branch calls `repo.split_once('/')`, returning `Err("invalid_repository")` when no slash is present. A model passing `{"repo": "ironclaw"}` always fails.  
**Fix:** (A) Remove `{ "required": ["repo"] }` from `anyOf`, keeping only `{ "required": ["owner", "repo"] }`; or (B) handle bare repo names without requiring a slash in `(None, Some(repo))`.

---

### 🟡 Medium — Security

**[SEC-1] `repo` field lacks pattern constraint allowing multi-slash/colon injection**  
`schema:24–28` — `repository` has `pattern: "^[^/\\s]+/[^/\\s]+$"`, but `repo` only has `minLength`/`maxLength`. Callers can pass `repo: "a/b/c"` or `repo: "foo:bar"`.  
**Fix:** Add `"pattern": "^[^/\\\\s]+(/?[^/\\\\s]+)?$"` to the `repo` field.

**[SEC-2] `validate_search_qualifier_value` permits `:` and `"` enabling GitHub query injection**  
`lib.rs:355–360` — Validator blocks whitespace/control chars but not `:`, `"`, `(`, `)`. These have special semantics in GitHub's query language. A value like `foo"label:bug"` (no space, passes validator) injects an unintended qualifier after URL-decoding.  
**Fix:** Extend to reject `:`, `"`, `(`, `)`.

---

### 🟡 Medium — Tests

**[TST-1] `wasm_guest_error_kind` has zero unit tests across all 9 arms**  
`runtime_adapters.rs:777–790` — New mapping function, zero coverage. Regression in any arm (invalid_ prefix, body_limit, timeout, egress_denied, forbidden, rate_limited, fallthrough) is invisible.  
**Fix:** `tests::wasm_guest_error_kind_maps_all_known_arms` exercising each arm.

**[TST-2] `search_query` owner-without-repo `Err("invalid_repository")` untested**  
`lib.rs:200` — `(Some(_), None)` arm has no test.  
**Fix:** `tests::search_query_rejects_owner_without_repo`.

**[TST-3] `search_query` `repository` field without slash untested**  
`lib.rs:183` — `split_once('/').is_none()` → `Err` path has no test.  
**Fix:** `tests::search_query_rejects_repository_without_slash`.

**[TST-4] `push_qualifier` whitespace-only filtering untested**  
`lib.rs:221–235` — Silently drops whitespace-only values but no test verifies this.  
**Fix:** `tests::push_qualifier_filters_whitespace_only_value`.

**[TST-5] `sanitize_host_error` fallthrough path (`"github_api_request_failed"`) untested**  
Fallthrough arm for unrecognised error strings has no test case.  
**Fix:** Add generic/unknown error assertion to `tests::sanitizes_host_egress_errors_without_leaking_details`.

---

### 🟡 Medium — Conventions

**[CON-1] Stringly-typed WASM error dispatch between guest and host**  
`runtime_adapters.rs:604,777` — `if error == "AuthRequired"` and `wasm_guest_error_kind()` match on literal strings. Per `AGENTS.md`: "Use strong types and enums over stringly-typed control flow when the shape is known." A casing change in the guest silently breaks host routing with no compile-time signal.  
**Fix:** Consider a small structured error type (`{"code": "auth_required"}`) deserialized by serde on the host side.

---

### 🟢 Low — Performance

**[PERF-1] Unconditional WASM asset re-materialization on every restore (every startup)**  
`extension_lifecycle.rs:85` — `materialize_available_extension()` now writes the ~152 KB WASM binary (and other assets) unconditionally for every installed extension on every process start. Pre-change, restore never called materialize.  
**Fix:** Skip if files already exist and manifest hash matches.

---

### 🟢 Low — Security

**[SEC-3] `starts_with("invalid_")` catch-all in `wasm_guest_error_kind` is runtime-wide**  
`runtime_adapters.rs:778` — Any future WASM guest returning a string starting with `invalid_` will be classified as `InputEncode`, masking server-side failures as user input errors.  
**Fix:** Enumerate specific known `invalid_*` strings rather than prefix matching.

---

### 🔵 Info — Conventions

**[CON-2] `state`/`issue_type` as `Option<String>` vs typed enums** — Low priority; follows pre-existing `sort`/`order` pattern in the same struct. An enum would catch invalid values at deserialization rather than in a separate validate function.

---

*False-positives filtered (2): "AuthRequired maps to OperationFailed" — handled at line 604 before `wasm_guest_error_kind`; "dead arms for github_api_forbidden/rate_limited" — `github_request()` now emits those exact strings.*

File: crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json (hunk 1 above)
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json]: **[BUG-2] Schema-runtime mismatch** — `{ "required": ["repo"] }` implies a caller may pass only `repo`. But `(None, Some(repo))` in `search_query()` calls `repo.split_once('/')` and returns `Err("invalid_repository")` when no slash is present. A model passing `{"repo": "ironclaw"}` always fails.

**Fix A:** Remove this variant; require `{ "required": ["owner", "repo"] }` instead.
**Fix B:** Handle bare repo names in the `(None, Some(repo))` branch without requiring a slash.

File: crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json [hunk 8]
```diff
@@ -10,6 +10,51 @@
...
+      "pattern": "^[^/\\s]+/[^/\\s]+$",
+      "description": "Optional owner/repo shorthand used to build a repo: qualifier when query is omitted or structured filters are supplied."
+    },
+    "owner": {
+      "type": "string",
+      "minLength": 1,
+      "maxLength": 100,
+      "description": "Optional repository owner. Must be paired with repo."
+    },
+    "repo": {
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/schemas/github/search_issues.input.v1.json:26]: **[SEC-1] `repo` field lacks format constraint** — `repository` has `pattern: "^[^/\\s]+/[^/\\s]+$"` but `repo` only has `minLength`/`maxLength`. Values like `a/b/c` (multiple slashes) or `foo:bar` pass schema validation and reach the WASM guest.

Fix: add `"pattern": "^[^/\\\\s]+(/?[^/\\\\s]+)?$"` to the `repo` field.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 9]
```diff
@@ -268,6 +352,27 @@ fn validate_path_segment(value: &str) -> bool {
             .any(|ch| ch.is_control() || ch.is_whitespace())
 }
 
+fn validate_search_qualifier_value(value: &str) -> bool {
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:366]: **[SEC-2] Qualifier value validator permits `:` and `"`** — GitHub search treats `:` as a qualifier separator and `"` as a phrase delimiter. A structured value containing these chars (no whitespace, passes this validator) can inject unintended qualifiers after GitHub decodes the URL-encoded query.

Fix: extend the validator to also reject `:`, `"`, `(`, `)`.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 10]
```diff
@@ -156,6 +167,73 @@ fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
...
+            .ok_or_else(|| "invalid_repository".to_string())?;
+        validate_repo(owner, repo)?;
+        parts.push(format!("repo:{owner}/{repo}"));
+    }
+    match (params.owner.as_deref(), params.repo.as_deref()) {
+        (Some(owner), Some(repo)) => {
+            validate_repo(owner, repo)?;
+            parts.push(format!("repo:{owner}/{repo}"));
+        }
+        (None, Some(repo)) => {
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:197]: **[BUG-2 continued]** — `repo.split_once('/')` errors when `repo` contains no slash (e.g., `"ironclaw"`), yet the schema allows `{ "required": ["repo"] }` alone. The field description says "owner/repo shorthand when owner is omitted" but the schema applies no pattern to enforce this. Either add a pattern constraint to the `repo` field in the schema, or handle bare repo names here without requiring a slash.

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 11]
```diff
@@ -156,6 +167,73 @@ fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
...
+            let (owner, repo) = repo
+                .split_once('/')
+                .ok_or_else(|| "invalid_repository".to_string())?;
+            validate_repo(owner, repo)?;
+            parts.push(format!("repo:{owner}/{repo}"));
+        }
+        (None, None) => {}
+        (Some(_), None) => return Err("invalid_repository".to_string()),
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:205]: **[TST-2] Missing test for owner-without-repo error path**
```rust
#[test]
fn search_query_rejects_owner_without_repo() {
    let result = search_query(&SearchIssuesParams {
        query: None, repository: None,
        owner: Some("nearai".to_string()), repo: None,
        author: None, assignee: None, involves: None,
        state: None, issue_type: None,
        page: None, limit: None, sort: None, order: None,
    });
    assert_eq!(result.unwrap_err(), "invalid_repository");
}
```

File: crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs [hunk 12]
```diff
@@ -156,6 +167,73 @@ fn search_issues(params: SearchIssuesParams) -> Result<String, String> {
...
+    }
+
+    if parts.is_empty() {
+        Err("invalid_query_empty".to_string())
+    } else {
+        Ok(parts.join(" "))
+    }
+}
+
+fn push_qualifier(
```
Comment (serrrfirat) [inline crates/ironclaw_first_party_extensions/assets/github/wasm-src/src/lib.rs:226]: **[TST-4] `push_qualifier` whitespace/control-char filtering untested** — This function silently drops whitespace-only values and errors on values with internal whitespace or control chars, but neither path has a test. Regressions here silently degrade search quality.

Add tests for: whitespace-only value (should return `Ok`, parts unchanged), internal-whitespace value (should return `Err`), control-char value (should return `Err`).

File: crates/ironclaw_host_runtime/src/services/runtime_adapters.rs (hunk 3 above)
Comment (serrrfirat) [inline crates/ironclaw_host_runtime/src/services/runtime_adapters.rs]: **[TST-1] `wasm_guest_error_kind` has zero unit tests** — All 9 logical arms (`invalid_*` prefix, `missing_invocation_context`, `unsupported_github_capability`, `github_api_body_limit`, `github_api_timeout`, `github_api_egress_denied`, `github_api_forbidden`, `github_api_rate_limited`, fallthrough) are untested. A regression in any arm is invisible.

Add `tests::wasm_guest_error_kind_maps_all_known_arms` exercising every arm.

File: crates/ironclaw_host_runtime/src/services/runtime_adapters.rs [hunk 13]
```diff
@@ -595,14 +595,20 @@ where
...
         }
     };
-    if execution.error.is_some() {
+    if let Some(error) = execution.error {
         account_or_release_failed_wasm_execution(
             request.governor,
             reservation.id,
             &execution.usage,
         )?;
+        if error == "AuthRequired" {
```
Comment (serrrfirat) [inline crates/ironclaw_host_runtime/src/services/runtime_adapters.rs]: **[CON-1] Stringly-typed WASM error dispatch** — `error == "AuthRequired"` couples the WASM guest string output to host routing logic. Per `AGENTS.md`: "Use strong types and enums over stringly-typed control flow when the shape is known." A casing change in `sanitize_host_error()` silently breaks auth routing.

Same concern applies to `starts_with("invalid_")` in `wasm_guest_error_kind` (line 778): any WASM guest emitting an error prefixed `invalid_` will be misclassified as `InputEncode` regardless of actual cause.

File: crates/ironclaw_reborn_composition/src/extension_lifecycle.rs [hunk 14]
```diff
@@ -81,6 +82,7 @@ pub(crate) async fn restore_extension_lifecycle_state(
         )?;
         let available = catalog.resolve(&package_ref)?;
         validate_restored_manifest_hash(&installation, available)?;
+        materialize_available_extension(filesystem.as_ref(), available).await?;
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/extension_lifecycle.rs:85]: **[PERF-1] Unconditional WASM asset re-materialization on every restore** — `materialize_available_extension()` writes the full WASM binary and all assets for every installed extension on every process start, unconditionally. Pre-change, restore never called materialize.

For N installed extensions the startup cost is now O(N × asset_bytes). Fix: skip if files are already present and match the manifest hash (already validated one line above).
